Update tests to use 1/0 not True/False addr type

AdamISZ opened the pull request:
Prior to this commit, several test functions were using
"True" to flag internal and "False" to flag external for
the HD branch for the wallet, but we now use 1/0, this
is changed. The tests passed both before and after this
change.

AdamISZ commented:
clarification: "we now use 1/0" refers to the `Wallet` class, see:

https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/9299191af17d6cd00a1ae70b6496f488b7788f87/jmclient/jmclient/wallet.py#L291-L292

kristapsk commented:
Shouldn't we export `ADDRESS_TYPE_EXTERNAL`/`ADDRESS_TYPE_INTERNAL` constants instead of using `0`/`1`?

AdamISZ commented:
It did occur to me but it didn't seem important enough, especially for a change only to tests.
I do agree, however unlikely it might be that these 0/1 values ever change.

But it will require additional code changes across the non-test codebase, and I will need to do considerably more testing work to ensure I haven't missed something.

AdamISZ commented:
Given the previous comment, @kristapsk are you OK with me merging it as is?

kristapsk commented:
Is this urgent to merge? I mean, tests pass anway. Or I'm missing something? Changing from 0/1 to constants everywhere will end up with a more readable code.

AdamISZ commented:
Well sure. "Tests pass anyway" - that was one of my concerns, we are using wrong arguments so I just want the tests to match correctly what the live code does; them passing is more valuable in the case where they don't call functions with different arguments than the live code.

A bunch more changes to the main code means more testing. I suppose I'll just have to do that.

AdamISZ commented:
See notes, updated as per discussion, and added a sanity-checking function in `BIP32Wallet` to make sure we only ever add entries to `self._index_cache` with valid branch keys.

Tests passing.

I'm currently trying to check as meticulously as I can that there are no other places in the code base where we explicitly reference the branch key (as 0 or 1, presumably) and if there are I'll update them accordingly.

AdamISZ commented:
Well, I already found a few. This will probably take a while ...

AdamISZ commented:
Latest push just adds a few more cases I found by carefully checking all calls of Wallet methods. As you can see only a couple of calls affected outside tests, but there are some internal changes in wallet.py which are more than enough reason to test and review this very carefully.

AdamISZ commented:
Despite my note above about caution, I do now consider this mergeable (it passes the test suite - which now hopefully represents the real code a little better! - any remaining details to address are very minor, I believe .. still needs a look over by someone else though).

kristapsk commented:
Just rebased on master, so that test suites passes with Bitcoin Core v0.21.

kristapsk reviewed (approved):
Looked at code, test suite passes, also did some manual wallet-tool tests, ACK 47b2a9cf0bc0a055c0d81e5f993ff075e91bdeee.
